src/lang/wasm.ts

```diff
@@ -120,26 +131,68 @@ const initialise = async () => {
 
 export const initPromise = initialise()
 
-export const rangeTypeFix = (ranges: number[][]): [number, number, number][] =>
-  ranges.map(([start, end, moduleId]) => [start, end, moduleId])
+const splitErrors = (
+  input: CompilationError[]
+): { errors: CompilationError[]; warnings: CompilationError[] } => {
+  let errors = []
+  let warnings = []
+  for (const i of input) {
+    if (i.severity === 'Warning') {
+      warnings.push(i)
+    } else {
+      errors.push(i)
+    }
+  }
+
+  return { errors, warnings }
+}
+
+export class ParseResult {
+  program: Node<Program> | null
+  errors: CompilationError[]
+  warnings: CompilationError[]
+
+  constructor(
+    program: Node<Program> | null,
+    errors: CompilationError[],
+    warnings: CompilationError[]
+  ) {
+    this.program = program
+    this.errors = errors
+    this.warnings = warnings
+  }
+
+  isOk(): boolean {
+    return !!this.program && this.errors.length === 0
+  }
```

Comment:
We can make a type predicate function if it's not a method. Could something like this work? I would need to play around with it.

```typescript
export interface SuccessParseResult extends ParseResult {
  program: Node<Program>
}

export function isOk(pResult: ParseResult): pResult is SuccessParseResult {
  return !!pResult.program && pResult.errors.length === 0
}
```

Reply:
Oh right, I tried this, it doesn't work because an assertion function is only allowed if it's result if always true. What we want to express is that the postcondition holds only if the result is true

Reply:
This is a predicate, not an assertion. Note the lack of `asserts` in the return type.

Reply:
Ha! Well this works, I won't to pretend to understand why...